File: packages/ui/server/utils/events.ts
import { event } from "~~/server/schema";
import { withTenantDb } from "~~/server/utils/db";
import type { EventType } from "~~/server/schema";
import { uuidv7 } from "uuidv7";

type EventScope = {
  project_id?: string;
  actor_id?: string;
};

type TxParam = Parameters<Parameters<typeof withTenantDb>[1]>[0];

export async function logEvent(
  organization_id: string,
  type: EventType,
  payload: Record<string, unknown>,
  system: boolean = false,
  scope: EventScope = {},
  tx?: TxParam,
) {
  const insert = (db: TxParam) =>
    db.insert(event).values({
      id: uuidv7(),
      organization_id,
      type,
      payload,
      system,
      ...scope,
      created_at: new Date(),
    });

  if (tx) {
    return insert(tx);
  }
  return withTenantDb([organization_id], insert);
}

File: packages/c-plane/src/services/events.rs
use chrono::Utc;
use sea_orm::{ActiveModelTrait, DatabaseTransaction, Set};
use uuid::Uuid;

use crate::{errors::AppError, models::entities::event};

pub async fn record(
    tx: &DatabaseTransaction,
    organization_id: Uuid,
    project_id: Uuid,
    event_type: &str,
    payload: serde_json::Value,
    actor_id: Uuid,
) -> Result<(), AppError> {
    event::ActiveModel {
        id: Set(Uuid::new_v4()),
        organization_id: Set(organization_id),
        event_type: Set(event_type.into()),
        payload: Set(payload),
        system: Set(false),
        project_id: Set(Some(project_id)),
        actor_id: Set(Some(actor_id)),
        created_at: Set(Utc::now().naive_utc()),
    }
    .insert(tx)
    .await?;
    Ok(())
}
